<a href="https://colab.research.google.com/github/atik-bd/MSc-BA/blob/main/Data%20preparation%20and%20Balancing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## APPLIED STATISTICS AND MACHINE LEARNING
## Exercise#1_ Atik Mahbub_ ID 20062447


# Data preparation (Breast Cancer dataset)

## Data Reading

In [ ]:
from pandas import read_csv,get_dummies,DataFrame
data1=read_csv('/content/Breast_Cancer_ 20062447.csv')

In [ ]:
data1.head()

Age   Race Marital Status T Stage  N Stage 6th Stage  \
0   68  White        Married       T1      N1       IIA   
1   50  White        Married       T2      N2      IIIA   
2   58  White       Divorced       T3      N3      IIIC   
3   58  White        Married       T1      N1       IIA   
4   47  White        Married       T2      N1       IIB   

               differentiate Grade   A Stage  Tumor Size Estrogen Status  \
0      Poorly differentiated     3  Regional           4        Positive   
1  Moderately differentiated     2  Regional          35        Positive   
2  Moderately differentiated     2  Regional          63        Positive   
3      Poorly differentiated     3  Regional          18        Positive   
4      Poorly differentiated     3  Regional          41        Positive   

  Progesterone Status  Regional Node Examined  Reginol Node Positive  \
0            Positive                      24                      1   
1            Positive                      14                      5   
2            Positive                      14                      7   
3            Positive                       2                      1   
4            Positive                       3                      1   

   Survival Months Status  
0               60  Alive  
1               62  Alive  
2               75  Alive  
3               84  Alive  
4               50  Alive

In [ ]:
data1.shape

(4024, 16)

In [ ]:
data1.describe()

Age   Tumor Size  Regional Node Examined  \
count  4024.000000  4024.000000             4024.000000   
mean     53.972167    30.473658               14.357107   
std       8.963134    21.119696                8.099675   
min      30.000000     1.000000                1.000000   
25%      47.000000    16.000000                9.000000   
50%      54.000000    25.000000               14.000000   
75%      61.000000    38.000000               19.000000   
max      69.000000   140.000000               61.000000   

       Reginol Node Positive  Survival Months  
count            4024.000000      4024.000000  
mean                4.158052        71.297962  
std                 5.109331        22.921430  
min                 1.000000         1.000000  
25%                 1.000000        56.000000  
50%                 2.000000        73.000000  
75%                 5.000000        90.000000  
max                46.000000       107.000000

In [ ]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Age                     4024 non-null   int64 
 1   Race                    4024 non-null   object
 2   Marital Status          4024 non-null   object
 3   T Stage                 4024 non-null   object
 4   N Stage                 4024 non-null   object
 5   6th Stage               4024 non-null   object
 6   differentiate           4024 non-null   object
 7   Grade                   4024 non-null   object
 8   A Stage                 4024 non-null   object
 9   Tumor Size              4024 non-null   int64 
 10  Estrogen Status         4024 non-null   object
 11  Progesterone Status     4024 non-null   object
 12  Regional Node Examined  4024 non-null   int64 
 13  Reginol Node Positive   4024 non-null   int64 
 14  Survival Months         4024 non-null   int64 
 15  Stat

## Data Cleaning
 In our dataset all 16 columns have equal number of datapoints, 4024 entries. So, data cleaning is not required here.

## Data Encoding

In [ ]:
data1['A Stage']
data1['Estrogen Status']
data1['Progesterone Status']
data1['Status']

0       Alive
1       Alive
2       Alive
3       Alive
4       Alive
        ...  
4019    Alive
4020    Alive
4021    Alive
4022    Alive
4023    Alive
Name: Status, Length: 4024, dtype: object

In [ ]:
data1['A Stage']=data1['A Stage'].map({'Regional':1,'Distant':0})
data1['Estrogen Status']=data1['Estrogen Status'].map({'Positive':1,'Negative':0})
data1['Progesterone Status']=data1['Progesterone Status'].map({'Positive':1,'Negative':0})
data1['Status']=data1['Status'].map({'Alive':1,'Dead':0})

In [ ]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Age                     4024 non-null   int64 
 1   Race                    4024 non-null   object
 2   Marital Status          4024 non-null   object
 3   T Stage                 4024 non-null   object
 4   N Stage                 4024 non-null   object
 5   6th Stage               4024 non-null   object
 6   differentiate           4024 non-null   object
 7   Grade                   4024 non-null   object
 8   A Stage                 4024 non-null   int64 
 9   Tumor Size              4024 non-null   int64 
 10  Estrogen Status         4024 non-null   int64 
 11  Progesterone Status     4024 non-null   int64 
 12  Regional Node Examined  4024 non-null   int64 
 13  Reginol Node Positive   4024 non-null   int64 
 14  Survival Months         4024 non-null   int64 
 15  Stat

In [ ]:
data2=get_dummies(data1,['Race','Marital Status','T Stage','N Stage','6th Stage','differentiate','Grade' ],dtype=int)

In [ ]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 37 columns):
 #   Column                                   Non-Null Count  Dtype
---  ------                                   --------------  -----
 0   Age                                      4024 non-null   int64
 1   A Stage                                  4024 non-null   int64
 2   Tumor Size                               4024 non-null   int64
 3   Estrogen Status                          4024 non-null   int64
 4   Progesterone Status                      4024 non-null   int64
 5   Regional Node Examined                   4024 non-null   int64
 6   Reginol Node Positive                    4024 non-null   int64
 7   Survival Months                          4024 non-null   int64
 8   Status                                   4024 non-null   int64
 9   Race_Black                               4024 non-null   int64
 10  Race_Other                               4024 non-null   int64
 11  Race

In [ ]:
data2.to_csv('Breast Cancer Data_Atik.csv')

## Defining X and Y

In [ ]:
y=data2['Status']
x=data2.drop('Status',axis=1)
print (x.shape)
print (y.shape)

(4024, 36)
(4024,)


## Scaling X

In [ ]:
from sklearn.preprocessing import StandardScaler
x_scaled=StandardScaler().fit_transform(x)
DataFrame(x_scaled)

0         1         2         3         4         5         6   \
0     1.565253  0.152963 -1.253661  0.267652  0.458107  1.190676 -0.618172   
1    -0.443222  0.152963  0.214345  0.267652  0.458107 -0.044095  0.164807   
2     0.449434  0.152963  1.540287  0.267652  0.458107 -0.044095  0.556296   
3     0.449434  0.152963 -0.590691  0.267652  0.458107 -1.525820 -0.618172   
4    -0.777968  0.152963  0.498475  0.267652  0.458107 -1.402343 -0.618172   
...        ...       ...       ...       ...       ...       ...       ...   
4019  0.895761  0.152963 -1.016886  0.267652  0.458107 -1.649297 -0.618172   
4020  0.226270  0.152963  0.735251  0.267652  0.458107 -0.044095  0.752041   
4021  1.565253  0.152963 -0.401271  0.267652 -2.182898 -0.414526 -0.226682   
4022  0.449434  0.152963  0.640541  0.267652  0.458107 -0.414526 -0.618172   
4023 -0.889550  0.152963 -0.022430  0.267652  0.458107 -0.908434 -0.422427   

            7         8         9   ...        26        27        28  \
0    -0.492961 -0.279201 -0.293927  ... -0.130123 -0.364531 -1.185437   
1    -0.405695 -0.279201 -0.293927  ... -0.130123 -0.364531  0.843571   
2     0.161530 -0.279201 -0.293927  ... -0.130123  2.743251  0.843571   
3     0.554224 -0.279201 -0.293927  ... -0.130123 -0.364531 -1.185437   
4    -0.929288 -0.279201 -0.293927  ... -0.130123 -0.364531 -1.185437   
...        ...       ...       ...  ...       ...       ...       ...   
4019 -0.972921 -0.279201  3.402205  ... -0.130123 -0.364531  0.843571   
4020 -0.100266 -0.279201 -0.293927  ... -0.130123 -0.364531  0.843571   
4021 -0.100266 -0.279201 -0.293927  ... -0.130123 -0.364531  0.843571   
4022  0.030632  3.581645 -0.293927  ... -0.130123 -0.364531  0.843571   
4023  1.252348 -0.279201 -0.293927  ... -0.130123 -0.364531  0.843571   

            29        30        31        32        33        34        35  
0     1.619247 -0.068877 -0.394955 -0.068877 -0.394955 -1.185437  1.619247  
1    -0.617571 -0.068877 -0.394955 -0.068877 -0.394955  0.843571 -0.617571  
2    -0.617571 -0.068877 -0.394955 -0.068877 -0.394955  0.843571 -0.617571  
3     1.619247 -0.068877 -0.394955 -0.068877 -0.394955 -1.185437  1.619247  
4     1.619247 -0.068877 -0.394955 -0.068877 -0.394955 -1.185437  1.619247  
...        ...       ...       ...       ...       ...       ...       ...  
4019 -0.617571 -0.068877 -0.394955 -0.068877 -0.394955  0.843571 -0.617571  
4020 -0.617571 -0.068877 -0.394955 -0.068877 -0.394955  0.843571 -0.617571  
4021 -0.617571 -0.068877 -0.394955 -0.068877 -0.394955  0.843571 -0.617571  
4022 -0.617571 -0.068877 -0.394955 -0.068877 -0.394955  0.843571 -0.617571  
4023 -0.617571 -0.068877 -0.394955 -0.068877 -0.394955  0.843571 -0.617571  

[4024 rows x 36 columns]

## Data Splitting

In [ ]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x_scaled,y,test_size=0.25, random_state=67)
print (x_train.shape)
print (x_test.shape)
print (y_train.shape)
print (y_test.shape)

(3018, 36)
(1006, 36)
(3018,)
(1006,)


## Data Balancing (for Training set)

In [ ]:
y_train.value_counts()

Status
1    2546
0     472
Name: count, dtype: int64

In [ ]:
from imblearn.over_sampling import SMOTE
x_train,y_train=SMOTE(random_state=103).fit_resample(x_train,y_train)

In [ ]:
y_train.value_counts()

Status
0    2546
1    2546
Name: count, dtype: int64